helpers/discordBot.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  SendChannelMessage,
+  SendDirectMessage,
+  SendLessonChannelMessage,
+  SendSubmissionNotification
+} from '../@types/discordBot'
+
+export enum IdType {
+  C0D3 = 'C0D3',
+  DISCORD = 'DISCORD'
+}
+
+export enum BotErrorType {
+  auth_error = 'auth_error',
+  validation_error = 'validation_error',
+  api_error = 'api_error'
+}
+
+const apiEndpoint = `${process.env.C0D3R_URL}/api`
+const messagesEndpoint = `${apiEndpoint}/messages`
+
+const headers = {
+  Authorization: `Bearer ${process.env.C0D3R_ACCESS_TOKEN}`,
+  'Content-Type': 'application/json'
+}
+
+export const sendChannelMessage: SendChannelMessage = async (
+  channelId,
+  message,
+  embed
+) =>
+  fetch(`${messagesEndpoint}/channel/${channelId}`, {
+    method: 'POST',
+    headers,
+    body: JSON.stringify({
+      message,
+      embed
+    })
+  }).then(r => r.json())
```

Comment:
Should probably add a `.catch()` on these fetches to handle unexpected errors that could interrupt the API call.

Could also maybe extract a helper function here with the `fetch` call since the parameters are the same for all the functions.

Reply:
Well there is actually another issue were we need to rememeber we are running serverless functions so unless you await every async call there will be no work finished after the user gets there response. 

But I don't think we really want our users waiting for a discord msg to go through so not awaiting is good in this case.

I think the real proper pattern we need to implement is if we want to catch all the errors is either auto respond with a msg received update and await that or just send it and know all error reporting is responsible via the receiver

Reply:
> Well there is actually another issue were we need to remember we are running serverless functions so unless you await every async call there will be no work finished after the user gets there response.

Yeah, when solving the issue with emails not being sent I believe part of the problem was that the promise was not being awaited. In this case the await should be put on the function that will be calling this, because the return of these functions are going to be 'Promise' even with the await.

For the error catching, in the case of these notifications I think they should be handled gracefully and not propagate up the call stack. Since these notifications are not crucial for the app they can be ignored and just be logged so us on the engineering team can look into it, but the response for the client should still be successful.

Reply:
My understanding is all these discord functions should be fire-and-forget. They send a command to the bot and that's it, this is to ensure that requests, of which these are called a part of, don't wait on response from discord as that is not necessary.

And so, keeping this in mind, like @tomrule007 pointed out, the serverless invocation that was handling a request may have already ended before discord responds.

Considering this, I'm not sure what a catch function would do here. The intended use is to call these without awaiting them or adding `.then`, if the caller needs the returned response then it would be their responsibility to take care of the error handling by either adding `.catch` then or if using async/await then surrounding it in a try catch block.

Does this make sense? if not then I'd need some suggestion as to what to do inside a `.catch` if a attach it to these.

Reply:
@ggwadera So is your suggestion then to await where they are called, and add a catch to these functions which silently logs the error on failure and lets the rest of the caller function run like usual ?

Reply:
Yes, having an await here won't make a difference, the function would still be asynchronous.

And because we don't really need the response here, `fetch` will throw an error on response statuses different from 2xx or 3xx. So to avoid having a try/catch everywhere we`re gonna call these functions, we can handle the exceptions here so the caller can function normally.

If an error is thrown here it would interrupt the caller and send an error response to the client. So if we had some more db operations to do after this for example, they would never be executed and we could end up with an invalid state on our database.

Reply:
It is my understanding as long as the promise propagates it's way back to the graphql server, the server will wait for resolve before sending a response to the user.

So the await would have to be on the 
`await sendSubmissionNotification(...)` in the `helpers/controllers/submissionController.ts  (line 69)` file

Then the catch statement could be anywere before or on this call. I talked with @JasirZaeem and told him I believe `.catch(console.error)` is the safest way to insure that error makes it to our log drain.

And since this response isn't important and we don't want to slow our users down unnecessarily I think we should follow the example from `js5/problem 7 image text extraction challenge` and have the `c0d3r` api always return a [Status 202 Accepted](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202) as soon as possible to prevent user slow down. We could also add a timeout to the fetch to reduce slow downs when the server is down.


Either way I am excited to get this discord bot going because I think it will be great for community engagement :)

Reply:
Hmmm good idea. The bot could reply immediately since we don't to need to wait for the Discord API call to finish it here. After the bot sends the response back it could the process the notification normally.

Detailed error logs could be done on the bot side too, here we could just leave a generic catch to avoid halting the execution.

Reply:
Yeah I am working on that, I will change the default behaviour of the bot to respond immediately without waiting on discord, while providing an option to get details of the message if needed, as these details can be used to perform tasks like replying to these messages, editing them reacting to them and so on.
So the default behaviour of the bot will be to respond without waiting on discord, but if requested it will wait for the message sending to complete and respond with the details of the message, this can then be used in the future to build more features.
Adding logging in the bot itself is a task in progress.

Reply:
That's because in most cases you would just fire and forget it, but, it does return a json body with any errors and (optionally) details of the message the was created at discord, so if a function wants to use any of these it can just acquire an object.

Reply:
- Calls to dicord functions are now awaited.
- Common fetch calls in discord functions have been moved into a separate function
- c0d3r will immediately respond to requests by default after https://github.com/garageScript/c0d3r/pull/11